### PR TITLE
Feature: New Thread to poll files.

### PR DESCRIPTION
**Pull request by simonnagl**
I closed the old pull request and reopen it here.
Sorry for that, I did something wrong with rebasing.

Added global polling. We now have a thread to poll files for becoming a specified state. Any other thread can register and erase files from the list of polled files.
The thread updates a timestamp which can be checked by the other thread.

### CODE REVIEW

**Comment by ktsaou** [comment]
nice! do you want a few suggestions?

**Comment by simonnagl** [comment]
Of course. I am always willing to learn. Please do so!

**Comment by ktsaou** [comment]
ok.

You are not obliged to follow any of these suggestions. Do as you see fit.

1. I suggest to simplify the poll API. Today, the rest of netdata has to keep track of timeval structures, know how to check if it has been updated, and free the final one. I suggest to have an API like this:

  - `void *poll_register(filename, type)` - much like you do today
  - `poll_is_updated(void *)` - check if it has been updated since the last call to `poll_is_updated()` or `poll_register()`.
  - `poll_unregister(void *)` - cleanup everything

  so, the caller will only need a `void *` for all operations.

2. As time passes, I understand netdata will become more complex. The poll API for example could be extended to monitor more external conditions (e.g. `inotify`, or something else). I know netdata today is far from this suggestion, but I tried hard to follow these rules with the dictionary and the registry and to follow them from now on.

  To make the code easily re-usable I try to apply some basic object oriented principles, like these:

  - the private globals of an object should go all together on a private struct. If these globals need to be read / manipulated by others, public functions should be created for this purpose.

  - every object has a constructor (`*_create()`)

  - every object has a destructor (`*_destroy()`)

  - every action that performs something on a object should be a function. For example you could have `poll_find()` to search for something.

  - locking is isolated (functions `*_lock()` and `*_unlock()`) so that we can experiment with different locking mechanisms without refactoring too much code (I find `futex` for example very appealing). I have already done this in dictionary, appconfig, registry, etc.

  - When locking is needed, I try to have at least 2 layers of functions. The internal `static *_nolock()` ones that do not handle locking and the higher level ones (the API ones) that handle everything. The code should finally look like this:

   ```c
my_public_api_call1() {
    my_api_lock();
    my_api_internal_function_nolock();
    my_api_unlock();
}
```

  So, between `*_lock()` and `*_unlock()`, only functions `*_nolock()` should appear. Also locking and unlocking should only be handled within the same function (no sharing of locks between functions).

  By the way, registry has 3 layers: a) nolock(), b) internal C-API that returns pointers/structures, c) web API that returns JSON. Only the third is exposed.

3. I know `poll` will never have too many entries to check, but as a general rule I try to avoid `strcmp()`. You can use a dictionary which is much simpler and a lot faster than a linked list (you will not need the linked list pointers and there are no strcmp() - the registry is full of dictionaries), or follow this simple rule:

  - store a hash of the string in the object: `uint32_t hash = simple_hash(string);`.
  - all `strcmp()` should be `if( hash == object->hash && !strcmp(string, object->string) )`. So we compare the hashes first and only if they match we make the function call to strcmp(). 90% of the netdata speed comes from this simple trick.

**Comment by ktsaou** [comment]
another one:

- you could eliminate the `pthread_kill()` if you add a timeout to `select()` for `update_every` seconds. `update_every` is the internal clock of netdata. So, when you register a poll, it will be executed at the next iteration (by default <= 1 sec). Your code is more precise, but personally I hate signals, so do as you see best.

**Comment by simonnagl** [comment]
Ok. Some comments:

1. This sounds good.
2. This sounds good too. I will try to hold on this principles. Of course we want to use established principles everywhere.
3. Of course. We should try to avoid CPU time wherever possible.

> I know poll will never have too many entries to check

If it will get bigger for some reason (much more than 100 file descriptors to monitor) we should switch to event based programming. I do not see a reason for this now. For our current needs `select` works much better.

> I hate signals
I never had problems with using signals. Is your hate founded somehow? Using signals to interrupt for registration reduces CPU consumption. Personally I like it very much.


I will refactor the code in the code and notify you when done. Thank you for your help.

**Comment by ktsaou** [comment]
> I never had problems with using signals. Is your hate founded somehow? Using signals to interrupt for registration reduces CPU consumption. Personally I like it very much.

Well, signals will interrupt the code at any point. This means that at random places you will have to check for return values that request a retry of the call. Also interrupts may happen inside locked code. This makes a lot of things a lot more complicated. So, I try to minimize their use to the absolute minimum.

**Comment by simonnagl** [comment]
Yes, of course. What you describe above is how to use signals the "right" way. This was also the reason for one of my last pull requests. Now the working threads are not handling signals anymore.
I will think more about your suggestion to use `update_every` for a timeout. Here I see two problems at the moment:
- Not predetermined when the polling starts. (The first value may be inaccurate).
- What about plugins setting their local update every less than `update_every`.

**Comment by ktsaou** [comment]
> Not predetermined when the polling starts. (The first value may be inaccurate).

Hm... we could have the first update happen with poll_register().

> What about plugins setting their local update every less than update_every.

Even if they do, the values that do not cross `update_every` although influence the calculations, they do not advance the charts. `rrdset_done()` is interpolating values at `update_every` boundaries. It handles all cases, from missing data collection points, to faster than expected collection. So even if a plugin collects values too fast, it will just waste its CPU cycles.

**Comment by ktsaou** [comment]
Keep in mind there is a global `update_every` that sets the minimum and each chart may have each own. However the later cannot be less than the global `update_every`.

**Comment by simonnagl** [comment]
This is the answer to https://github.com/firehol/netdata/issues/407#issuecomment-221477858. I put it here to not split discussion.

I read the manpage of inotify. It supports quite more information as simple polling. Currently I am not sure if we want to mix these two calls into one API.

In #407 inotify can detect the following changes:
- creted file.
- deleted file.
- moved file.
- changed permissions of a file.

Of course we could use it to only monitor if a event occurred on a file, but especially for directories we should implement some way to query which event happend. (See the bullet  list above).
@ktsaou, If you want to make a suggestion for a API please do so. I am busy this week, so I maybe can make one at the weekend.

Regardless how we design the API we should implement the functionality on the same thread. We simply poll for the inotify file descriptor and read the inotify queue if it became readable.

**Comment by ktsaou** [comment]
ok. I haven't looked to it in detail.
There is no hurry. inotify support was just FYI.

**Comment by simonnagl** [comment]
I rebased a new version.

The major reason why I did not remove signals:
- Closing a file descriptor while select is running is unspecified.

If we do not use signal to interrupt select we would have to block a thread that wants to register a file until select unblocks because of timeout. I do not think we should do that.

Tomorrow I will review my comments and update them if I find something to enhance. Can you please make suggestions for enhancements if you have some? Thank you!

**Comment by simonnagl** [comment]
I've done some comment changes. @ktsaou, it's your turn.

File: src/poll.c [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+int poll_time_update_nolock(struct timeval *tv) {
+	int retv = gettimeofday(tv, NULL);
+	if(retv != 0) {
+		error("Could not get current time");
+	}
+	return retv;
+}
+
+// if bigger is less than lower we return 0.
+unsigned long long poll_time_difference_nolock(struct timeval *bigger, struct timeval *lower) {
```
Comment (simonnagl) [inline src/poll.c]: I do not like this name. Anyways I do not find a better one.

File: src/poll.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#define POLLIN 1
+#define POLLOUT 2
+#define POLLERR 3
+
+/**
+ * @brief Thread loop.
+ *
+ * This must only be called once by the main thread.
+ *
+ * @param ptr TODO
```
Comment (simonnagl) [inline src/poll.h]: There I need your help.

File: src/poll.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * @return TODO
```
Comment (simonnagl) [inline src/poll.h]: There I need your help too.

File: src/poll.c [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+	return retv;
+}
+
+unsigned long long poll_occured(void *poll_descriptor) {
+	struct poll_check *p_check = poll_descriptor;
+	unsigned long long retv;
+
+	poll_file_list_lock();
+	retv = poll_time_difference_nolock(&p_check->poll_file->tv, &p_check->tv);
```
Comment (ktsaou) [inline src/poll.c]: I think no locking is needed here. The whole idea of locking is to protect the linked list. We could assume that for each poll entry, the owner is only one thread. So, there is no way to delete an entry while the caller access it.

If there is the possibility of multiple threads accessing the same poll, even the `void *` we return might be invalid.

File: src/poll.h (hunk 2 above)
Comment (ktsaou) [inline src/poll.h]: for what?

File: src/poll.h (hunk 3 above)
Comment (ktsaou) [inline src/poll.h]: for what?

File: src/poll.h (hunk 2 above)
Comment (simonnagl) [inline src/poll.h]: For the description what's the parameter ptr for pthreads.

File: src/poll.h (hunk 3 above)
Comment (simonnagl) [inline src/poll.h]: For the description how to use return value for pthreads.

Or should we just remove the detailed description from this methods because both values are not used.

File: src/poll.c (hunk 4 above)
Comment (simonnagl) [inline src/poll.c]: Yes, `poll_occured()` should not be called with the same pointer from different threads. But `poll_time_difference_nolock()` reads from `p_check->poll_file->tv` which is also in the linked list. For the unlikely case that we read it while the polling loop updates it we might get a dirty read.
I did not lock the update of `p_check->tv` (see below) for the same reason you described.

File: src/poll.h (hunk 3 above)
Comment (ktsaou) [inline src/poll.h]: ok, they are not currently used, so it would be ok to remove them.

File: src/poll.c (hunk 4 above)
Comment (ktsaou) [inline src/poll.c]: I think "dirty" here is not something bad. Either the old or the new value will be read. There is no unpredictable behaviour.

Generally, adding a lock at this point, means that every thread that will make this call to find out if the poll has been updated, a lock/unlock will be executed. Every poll, ever second.

If you don't trust the way the value is updated, I suggest to have another copy of it and update it ourselves, to eliminate this lock.

Comment (simonnagl) [inline src/poll.c]: "dirty" is something bad. T
There is no guarantee that the write `gettimeofday` is atomic. Even there is no guarantee that a write to the underlying data types of timeval are atomic.

Having a copy  will not solve the problem. We always will have to write to the copy in another thread than reading it. If we want to do it in a save way we need to synchronize this calls.

When thinking about removing the lock here we should think what could happen. We could have a dirty read. This means we read data we did not write. This can lead to a wrong return value of `poll_occured`. This can only happen when the file got polled and we check the polling while updating the timeval to check. There can be two errors:

1. We miss the poll in this call. The next call will get it.
2. We get a wrong return value. For now this seems ok to.

Because of the little probability this might happen and the little impact if it happens I would remove the lock here.

I do not know how much performance locking costs....

Comment (ktsaou) [inline src/poll.c]: Can you describe the worst case scenario?

The worst I can think of, is `tv.tv_sec` to be updated but `tv_tv_usec` not  be updated. So the microseconds of the old value would appear as the microseconds of the new.

To solve this, we could keep a timestamp as `unsigned long long last_updated = tv.tv_sec * 1000000ULL + tv.tv_usec`. This can be updated inside the locked code of the poll.

Since this is now a single variable we can trust that either we read the old value or the new value, as a whole.

Comment (ktsaou) [inline src/poll.c]: Locks are performance killers. This is why many high level languages (like javascript, python, Go, etc) are single threaded: they provide an illusion of threads by using async-I/O and so they guarantee no real locks are needed.

Also many NoSQL dbs are single threaded for the same reason (see redis).

Lockless multi-threaded programming exists, but requires special care (check http://liblfds.org/). There are even special instructions on the CPUs to allow lockless execution.

The linux kernel itself has the [RCU](https://lwn.net/Articles/262464/), which is a special linked list that permits multiple readers and one writer concurrently. But they are have special code that knows how the underlying processor works and precisely allows this to happen.

So, locks are bad. Actually, no matter how good your code is single threaded, multi-threading with locks can really destroy it.

Comment (ktsaou) [inline src/poll.c]: And... do you remember the statistics of the registry?

The registry with locks, but running single threaded is able to do 100.000 queries per second per core.

It goes down to 5.000 queries per second per core with locks and multiple threads running. It is 20 times slower just because of locks.

Comment (simonnagl) [inline src/poll.c]: > Since this is now a single variable we can trust that either we read the old value or the new value, as a whole.

Are you sure? `usigned long long` for example will only be updated atomically if it requires

1. less memory than the system register size.
2. the compiler aligns it right.

While 2 should not be a problem for us 1 can happen. `long long` is normally 8 Bytes long. 32 Bit systems have to small registers.

This is something RCU tries to prevent writing somewhere else and then swapping the pointer which is atomic.

I tried to implement something similar. I pushed my changes.

File: src/poll.c [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+		error("Could not get current time");
+	}
+	return retv;
+}
+
+// Updates p->tv to current time.
+// Ensure this method is only called in one thread.
+// Ensure there are no writes to p in other threads.
+// Other writes are permitted without locks.
+int poll_file_time_update_readsave_nolock(struct poll_file *p) {
```
Comment (simonnagl) [inline src/poll.c]: This function ensures clean reads on p->tv without locks on any thread.

**Comment by ktsaou** [comment]
well, I lost the whole day trying to make netdata go above 5.000 queries/s. I managed to get it up to 10.000/s but crazy things happen. The listening socket becomes unresponsive (even closing and re-opening it does not work) and several threads consume 100% CPU within `select()` (I core dumped it to see it and yes, they run withing glibc).

After several hours I figured out that `select()` works only for file descriptors <1024. It is on the man page. To `select()` a file descriptor above that, we have to use `poll()`. I learned it the hard way!

I will try to adapt the web server tomorrow to use `poll()`. This means that this code will also break if netdata is having more that 1000 sockets open (~26 files are always open for data collection).

**Comment by simonnagl** [comment]
Yes, select only works for <1024 file descriptors. Damfool that you had to learn it the hard way! Poll has the same problem like select. It does not scale good. Both functions copy the file descriptor set from userspace to kernelspace on each call.

The faster alternative, which I would choose in the web server is `epoll`, which scales much better. The problem with it is portability. `epoll` is linux specific.

> This means that this code will also break if netdata is having more that 1000 sockets open.

You are right. I did not think about this. I used `select` because of portability. I will switch the code to use `poll`.

**Comment by ktsaou** [comment]
Nice discussion about alternatives: http://www.wangafu.net/~nickm/libevent-book/01_intro.html

Probably this is what we need: http://libevent.org/

**Comment by simonnagl** [comment]
Thank you for the links. Using a library for this seams appropriate for me.

http://software.schmorp.de/pkg/libev.html is a alternative to libevent.

**Comment by ktsaou** [comment]
ok, for the moment I have converted to multi-threaded web server to `poll()` which does not suffer from such a hard limit. With this I have managed to get 9.000 queries/s on a single core machine with 2.000 connected clients.

I have also added a single threaded web server but its performance is really killed by the loops required by the use of `select()` (and probably `poll()`). I will leave it in the code, until we use a library for events.

I suggest to use `poll()` in this module also. It really simple, almost a drop-in replacement for select.
I will make a PR in a few minutes, so you can check the code.

**Comment by ktsaou** [comment]
libev seems sweet!

**Comment by simonnagl** [comment]
Switched to poll. This requires to use arrays. This is not yet tested very well. I will do this in the next days and ping you. Please make your comments as always. You know I am willing to learn.

**Comment by simonnagl** [comment]
Apart from this I took a closer look to the two event libraries.

If we want to use one of it in the feature I suggest to use `libevent`. If we implement a event loop we will have to add events from other threads and notify the threads when the event occurred. The API of `libivent` manages the locking itself, while `libev` requires additional code. As you can see here: http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#THREAD_LOCKING_EXAMPLE

Anyways I am not sure if we need to use them in the future.
- The server threads only poll for one or two file descriptors at once. The main problem with `poll` and `select` is that you have go through the whole set of file descriptors and reload them into the kernel space. With the approach we have there now using `epoll` might not be faster.
- This module also might not poll for dozen of files.
- The libraries do not support all features of `inotify`.
- Might we gain something with having timer events, synchronus signal handling or forks?

**Comment by ktsaou** [comment]
nice work!
Is this mergeable?

> Anyways I am not sure if we need to use them in the future.

me neither...

This might be related though: https://github.com/nghttp2/nghttp2
Browser support for HTTP/2: http://caniuse.com/#feat=http2
HTTP/2 in netdata could be nice...


> Might we gain something with having timer events, synchronus signal handling or forks?

Can you elaborate a bit? For doing what?

**Comment by simonnagl** [comment]
> Might we gain something with having timer events, synchronus signal handling or forks?

For doing what? That is the question. If yes this might be a reason to use a library for it. At the moment I can not figure out any reason to do so.

Thank you for the links. I will look at them later.

This might be mergeable. It is not well tested. Please give me some time to test it. I will ping you when done. I do not want to merge barely tested code.

**Comment by simonnagl** [comment]
Done. This now is mergeable. I fixed some memory leaks and made the code saver for wrong calls. I removed the test code from `diskstats`.  

Tested with multiple file descriptors, registration from multiple threads, invalid files (and files that can not be open).

File: src/poll.c [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+
+// ---------------------------------------------------------------------------
+// internal methods
+int poll_time_update_nolock(struct timeval *tv) {
+	int retv = gettimeofday(tv, NULL);
+	if(retv != 0) {
+		error("Could not get current time");
+	}
+	return retv;
+}
```
Comment (ktsaou) [inline src/poll.c]: this is almost the same with:

```c
#define poll_time_update_nolock(tv) gettimeofday(tv, NULL)
```

the only difference is an error log.

File: src/poll.c [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+// Updates p->tv to current time.
+// This is not thread save.
+// concurrent reads are save. Concurrent writes need to be syncronized.
```
Comment (ktsaou) [inline src/poll.c]: save = safe

File: src/poll.c [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+int poll_data_time_update_readsave_nolock(struct poll_data *d) {
```
Comment (ktsaou) [inline src/poll.c]: poll_data_time_update_readsa**f**e_nolock

File: src/poll.c [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+	buff->tv_usec = d->tv->tv_sec;
+	int retv = poll_time_update_nolock(buff);
+
+	// Switch pointers.
+	struct timeval *tv = d->tv;
+	d->tv = buff;
+	buff = tv;
+
+	return retv;
+}
```
Comment (ktsaou) [inline src/poll.c]: oops! this will not work if this function is called from multiple threads concurrently.
The right way is to have 2 `struct timeval` allocated inside `struct poll_data` and also a pointer to the latest one.

File: src/poll.c [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+		return 0;
+	}
+}
+
+int poll_array_search_nolock(char *path, int type) {
+	uint32_t hash = simple_hash(path);
+	int i;
+	for(i = 0; i < poll_num; i++)
+		if(poll_fd_array[i].fd < 0 && poll_data_array[i].path_hash == hash && poll_fd_array[i].events == type && \
+				strncmp(poll_data_array[i].path, path, strlen(poll_data_array[i].path)) == 0)
```
Comment (ktsaou) [inline src/poll.c]: why `strncmp()`? Are you searching an equal prefix (strncmp) or exact match (strcmp)?

Comment (ktsaou) [inline src/poll.c]: Also ` \ ` is not needed in C.

File: src/poll.c [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+
+int poll_array_find_next_deprecated_nolock(int index) {
+	for(index++; index<poll_num; index++)
+		if(poll_fd_array[index].fd < 0)
+			return index;
+	return -1;
+}
+
+void poll_array_remove_nolock(int index) {
+	if(poll_fd_array[index].fd >= 0)
```
Comment (ktsaou) [inline src/poll.c]: insert:

```c
if(index >= poll_num) return;
```

File: src/poll.c [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+		if(close(poll_fd_array[index].fd) != 0)
+			error("Failed to proper close file descriptor %d", poll_fd_array[index].fd);
+	poll_fd_array[index].fd = -1;	
+	if(poll_first_deprecated < index)
+		poll_first_deprecated = index;
+}
```
Comment (ktsaou) [inline src/poll.c]: I think additional error logging in this function could ensure we have not messed everything up...

File: src/poll.c [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+			return i;
+	return -1;
+}
+
+int poll_array_find_next_deprecated_nolock(int index) {
+	for(index++; index<poll_num; index++)
+		if(poll_fd_array[index].fd < 0)
+			return index;
+	return -1;
+}
```
Comment (ktsaou) [inline src/poll.c]: what if there is a free slot before `index`?

This function a called when we need to allocate a slot with `index = poll_first_deprecated`. So check this timeline:

- at `t` slot 3 (`index = 2`) if free, so `poll_first_deprecated = 2`
- at `t+1` slot 2 (`index = 1`) is removed
- at `t+2` a new poll is requested, `poll_array_add_nolock()` calls this functions, but does not find slot 2 available.

File: src/poll.c [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+	poll_fd_array[index].fd = -1;	
+	if(poll_first_deprecated < index)
+		poll_first_deprecated = index;
+}
+
+int poll_array_add_nolock(char *path, int type) {
+	int index;
+
+	if(poll_first_deprecated >= 0) {
+		index = poll_first_deprecated;
```
Comment (ktsaou) [inline src/poll.c]: what if `poll_first_deprecated == -1`?
A SIGSEGV...

**Comment by ktsaou** [comment]
@simonnagl sorry... this is not ready. I stopped reviewing in the middle of it.

I understand this `pollfd` array increased complexity.

I would suggest to remove all the `deprecated` logic. This is an over-optimization and does not seem to work well. There are techniques to avoid the loops in the array, but let's say these are not required for this module.

So, simplify it a bit and let me know when you are done to review it again.

File: src/poll.c (hunk 8 above)
Comment (simonnagl) [inline src/poll.c]: Thank you.

File: src/poll.c (hunk 7 above)
Comment (simonnagl) [inline src/poll.c]: Thank you.

File: src/poll.c (hunk 9 above)
Comment (simonnagl) [inline src/poll.c]: You are right. I have done it like this because it only should be called from the polling thread. As described on top of the function.
If we want to switch it as you described we might get a dirty read.

Thread 1: `buff->tv_sec = d->tv->tv_sec;`
Thread 2: `d->tv = buff;`
Thread 1: `buff->tv_usec = d->tv->tv_usec;`

This might not be severe but I do not like it anyways.

File: src/poll.c [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+	if(!buff) {
+		buff = malloc(sizeof(struct timeval));
+		if(!buff) {
+			error("Could not allocate buffer for poll_data_time_update_readsave_nolock()");
+			return -1;
+		}
+	}
+
+	buff->tv_sec = d->tv->tv_sec;
+	buff->tv_usec = d->tv->tv_sec;
```
Comment (simonnagl) [inline src/poll.c]: Typo: I will correct it:
buff->tv_usec = d->tv->tv_**u**sec

File: src/poll.c (hunk 6 above)
Comment (simonnagl) [inline src/poll.c]: Ok. I am wiling to change it. Should I log an error at the call of `poll_time_update_nolock`?

File: src/poll.c (hunk 10 above)
Comment (simonnagl) [inline src/poll.c]: You are right. Sorry for that mistake.

File: src/poll.c (hunk 13 above)
Comment (simonnagl) [inline src/poll.c]: in `poll_array_remove_nolock()`
```
if(poll_first_deprecated < index)
    poll_first_deprecated = index;
```

File: src/poll.c (hunk 14 above)
Comment (simonnagl) [inline src/poll.c]: The line above: `poll_first_deprecated >= 0`.
If `poll_first_deprecated == -1` we do have to increase the array.

**Comment by simonnagl** [comment]
`this is not ready`
Have you tested it or you just read the code?
I see this code maybe is over-optimized. Definitely it is to difficult to read. I will remove the deprecated logic and simply loop over the array finding the next free position.

Please answer my comment at `poll_data_time_update_readsave_nolock`. I do not think there is a save way, and nor need for updating the timeval in different threads without locking.

**Comment by ktsaou** [comment]
> Have you tested it or you just read the code?

No, I haven't tested it yet. I may be over-protective with our users. I hate hunting bugs. So, I prefer to think twice before merging something (this goes to my code too). Of course I will test it if I can't see something wrong in the code.

Don't be frustrated. There is no reason. I may have judged wrong. I am human too...

> Please answer my comment at poll_data_time_update_readsave_nolock. I do not think there is a save way, and nor need for updating the timeval in different threads without locking.

For some reason this is has become a problem, but I think the solution is simple:

1. The polling thread should have 2 `struct timeval` buffers and a pointer to the latest one updated. Will first write data to the **unused** buffer and then switch the pointer to it.

2. The callers should hold a `struct timeval` buffer with their latest time data. Will only compare their buffer with the one pointed by the pointer of the polling thread, and after every comparison they will copy the polling thread buffer to their own. They should use the pointer of the polling thread, just once (this is important).

If we do it like that, I don't see the issue.

**Comment by ktsaou** [comment]
@simonnagl do you want me to finish this?

**Comment by simonnagl** [comment]
Sorry for the late reply. If you want you can do this. I'm afraid that I will be busy the next 5 weeks. If you do not care to have this open such a long time I will finish it afterwards. If you want to finish it please do so. I then will come back and pick another enhancement to work on if you do not care!

PS: This has nothing to do with frustration. You are making a perfect job and your comments are reasonable!

**Comment by ktsaou** [comment]
ok. I think we can wait for you. This is your baby...

**Comment by simonnagl** [comment]
@ktsaou:

I finally cleaned up the code. But it is still not ready. There is one point I want to discuss.

For that commit I made `poll_occured` blocking again. This is what we should change.

As we made clear in former discussion we need to switch the the timeval-pointer when updating `poll_data_array[i].tv`.

Your approach would be having two `struct timeval` for every `poll_data_array`.
My approach would be having one buffer_tv which could be a static property of a function or a property next to `struct poll_data`.

For performance I do not see a difference in both approaches.
For code quality I think mine would be more straight-forward but this may be matter of taste.
Mine uses less memory.

Do you agree?

**Comment by ktsaou** [comment]
welcome back!

Give me a day to review this again.

Keep in mind I have used CPU atomic operations when they are available (check for example: https://github.com/firehol/netdata/blob/3e2e1707e21eb35778b4ea53b7d5834b9b4ce142/src/global_statistics.c#L59-L71). We could use them to swap the pointers here too.

Also, I have changed diskstats to include space statistics for all disks (even remote or virtual disks, much like `df` does) and made several improvements to mountinfo based on the source code of `coreutils`.

So, give me a day, to review the comments and code.

nice to have you back!

File: src/poll.c [hunk 16]
```diff
@@ -0,0 +1,356 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <sys/poll.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include "common.h"
+#include "poll.h"
+#include "log.h"
+
```
Comment (ktsaou) [inline src/poll.c]: include files are now centrally managed in `common.h`, so that all files in netdata include only `common.h`.

File: src/poll.c [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+}
+
+void poll_check_free(struct poll_check *p_check) {
+	free(p_check);
+}
+
+// Interrupts select if the process is currently blocking.
+void poll_interrupt() {
+	if(poll_thread) {
+		pthread_kill(poll_thread, SIGUSR2);
```
Comment (ktsaou) [inline src/poll.c]: SIGUSR2 is now used for reloading health monitoring.

File: src/poll.c [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+	poll_fd_array[index].fd = -1;	
+}
+
+int poll_array_add_nolock(char *path, int type) {
+	int index;
+
+	index = poll_array_find_first_deprecated_nolock();
+	if(index >= 0) {
+		poll_num++;
+		poll_fd_array = realloc(poll_fd_array, sizeof(struct pollfd) * poll_num);
```
Comment (ktsaou) [inline src/poll.c]: also memory allocation functions now use the same ending with `z`. This takes care of failures, so you don't need to check for NULL responses.

i.e. `mallocz()`, `callocz()`, `reallocz()`, `strdupz()`, `freez()`.

**Comment by ktsaou** [comment]
> For performance I do not see a difference in both approaches.
> For code quality I think mine would be more straight-forward but this may be matter of taste.
> Mine uses less memory.
> 
> Do you agree?

@simonnagl do as you see fit. I don't see much of a difference, so pick the one you believe is best.

I have made a few comments. netdata has changed since May, so a few changes are needed to comply with the rest of the code.

The use of SIGUSR2 is probably a problem. I have restricted SIGUSR2 to the health monitoring thread.
Instead of using another signal, I suggest to do what I have recently seen on another project and found it brilliant: a pipe. So, the poll thread will open a pipe, where one end goes in the `poll()` call and the other is used by other threads to write a byte, just to wake up the polling thread. Much cleaner approach, avoiding signals entirely.

**Comment by simonnagl** [comment]
Thank you for the hints.
I will take care.

SIGUSR2: I will check how you implemented it for health monitoring. If it is just used by other threads, not processes there should be no problem. We can send SIGUSR2 from threads either to the health monitoring thread or to the polling thread.
The solution with the pipe sounds not bead but I am not sure if we can interrupt poll with this approach. Maybe using signals is cleaner.
Another question is if SIGUSR1 is already used by the application.

As I said I will take a closer look and implement the approach which looks best for me. Of course I want you to review it. I won't be disappointed!

I definitely  will use atomic operations if available and use locking if not. This API sounds great.

Implementing this might take some weeks. Pleas be patient. I still have much private things to do but working on this project is really fun for me!

**Comment by ktsaou** [comment]
ok, nice.

Keep in mind that the pipe will not "interrupt" `poll()`. One of the file descriptors passed in the array of descriptors to `poll()`, will be the reading side of the pipe. To my understanding this is the cleanest and most reliable way to do it. There will not be any interrupt, there will be no timeout. `poll()` will unblock the same way it does if a monitored file changed, but you will know it is the pipe that did that and just update the structures needed for the next iteration.

Think of it and do as you see fit.

**Comment by simonnagl** [comment]
Hello again. Yesterday I tried to implement the recommended changes. It think it is not as easy as it seams to be. If you agree I will give up on this feature and help you out somewhere else.

First of all I will look at the other open PR. Is that OK for you?

**Comment by ktsaou** [comment]
There is nothing that depends on this currently.
So, of course. Do as you see please...

**Comment by simonnagl** [comment]
This will not be merged. @ktsaou explains why:

> To have a cross-platform netdata, we will remove all POSIX/Linux/*BSD related calls and replace network I/O, disk I/O, threads and locks, with libuv ( #2637 ).
> 
> Since libuv is running on all platforms to support node.js, netdata will also be able to be compiled on all known platforms out there, including MS Windows ( #92 ), and will actually use the best polling mechanism available on each platform (ie. epoll() on Linux, kqueue() on FreeBSD, etc).
> 
> @vlvkobal has already started learning libuv, so I believe soon enough we will have progress on this.
> 
> It is important for netdata to run in abnormal system situations. For example, netdata /proc plugin is able to work perfectly when the system disks have crashed! @Ferroin has verified this recently. This means we will never use a global polling mechanism for everything, since something blocking there, will block everything. We will always use multiple event loops, grouped by nature, according to the work of each module.

I will close this.
